### ISSUE TITLE
storage,sql: find orthogonality in `CREATE SOURCE` syntax

### ISSUE THREAD

**Issue by benesch** [MaterializeInc/materialize#11957]
There are several moments where we lose orthogonality in `CREATE SOURCE`. Whenever you can't freely mix-and-match connectors, formats, and envelopes, we've got an orthogonality problem.

We have a much better handle on orthogonality now than we did a year ago, thanks to a lot of work from @umanwizard and @petrosagg. We should take the opportunity with Materialize Platform to see if we can gently redesign the `CREATE SOURCE` syntax to match the reality of what is allowed. 

For example, does `ENVELOPE DEBEZIUM` make sense? It only works with Kafka and expects a very specific Avro schema. 

@guswynn, can you flesh out with some details?

**Comment by benesch** [comment]
It seems to me that a good starting point would be tackling any of the many TODOs in `plan_create_source`:

* https://github.com/MaterializeInc/materialize/blob/d90add28a6f2b7ea2e05eb424f2874ab75b17921/src/sql/src/plan/statement/ddl.rs#L623
* https://github.com/MaterializeInc/materialize/blob/d90add28a6f2b7ea2e05eb424f2874ab75b17921/src/sql/src/plan/statement/ddl.rs#L630
* https://github.com/MaterializeInc/materialize/blob/d90add28a6f2b7ea2e05eb424f2874ab75b17921/src/sql/src/plan/statement/ddl.rs#L638
* https://github.com/MaterializeInc/materialize/blob/d90add28a6f2b7ea2e05eb424f2874ab75b17921/src/sql/src/plan/statement/ddl.rs#L787
* https://github.com/MaterializeInc/materialize/blob/d90add28a6f2b7ea2e05eb424f2874ab75b17921/src/sql/src/plan/statement/ddl.rs#L795
* https://github.com/MaterializeInc/materialize/blob/d90add28a6f2b7ea2e05eb424f2874ab75b17921/src/sql/src/plan/statement/ddl.rs#L809
* https://github.com/MaterializeInc/materialize/blob/d90add28a6f2b7ea2e05eb424f2874ab75b17921/src/sql/src/plan/statement/ddl.rs#L824-L825

**Comment by petrosagg** [comment]
This may be a hot take but I think we should stop using the term orthogonality because it heavily implies that most options are compatible with most options when that is not the case. It sounds much more natural to me to think in terms of composable primitives and a source pipeline that transforms typed collections.

When defining a source you have a few primitives at hand (bare sources) where each source documents what kind of schema it outputs. Users also have some transformations available to them that they can apply (e.g decoding, envelope logic, etc). Documentation can guide them on what each transformation expects as an input and what it produces at its output and provide a very similar experience as composing functions when programming.

For example I think it should be possible for a user to create a table with bytes and then apply a decoding transformation on top of it and have that as a source. If we do plan to have the long tail of sources handled by INSERTs into a table this might end up being necessary.

**Comment by benesch** [comment]
> This may be a hot take but I think we should stop using the term orthogonality because it heavily implies that most options are compatible with most options when that is not the case. It sounds much more natural to me to think in terms of composable primitives and a source pipeline that transforms typed collections.

I love this take. I'll rename this issue to use the term "composability" instead.

**Comment by frankmcsherry** [comment]
Riffing off of https://github.com/MaterializeInc/materialize/issues/11957#issuecomment-1108310988 I wanted to record what I thought was a good idea! Maybe others do too, idk.

We have a few different flavors of "thing" we are manipulating in the source pipeline. I think of them as three:
1. Raw source data, specific to the source type, without generally reusable properties.
2. Timestamped data with a frontier, representing events at times and a guarantee that future events will have other times.
3. Differential data with a frontier, representing a collection of something with insertions and deletions and all that.

Essentially all of our sources start as (1.). Going to (2.) is what I think of as "reclocking": the first moment at which we have a system timestamp associated with input data. We have a pile of operations that go from (2.) to (2.) e.g. decoding, perhaps some upsert pre-work, stuff like that. I think of going from (2.) to (3.) as our "envelope" work.

Each of these transitions could involve multiple inputs and multiple outputs. For example, decoding might want timestamped schema registry data as an input, to make the decoding deterministic. Enveloping (?) might want a timestamped data and transaction metadata to produce timestamped complete transactions.

So, following the comment, we have a pile of actions we can apply to things that may have the right shape; to a first approximation, which of (1.), (2.), and (3.) are they, but then in addition do we have the right number of inputs, and do they each have the right types of record?

---

We discussed this a bit in DMs, and there was not clear consensus that this framing was best. I'm not sure anyone thought it was wrong, so much as potentially limiting? (( hop in if I've misrepresented. )) Example concerns were: "will we be able to respect the timestamp properties" and "might we want to bring together data with unaligned timestamps and emit a record of what we did".

**Comment by nmeagan11** [comment]
> For example, does ENVELOPE DEBEZIUM make sense? It only works with Kafka and expects a very specific Avro schema.

@benesch, out of curiosity, why do you think this is the case? Isn't `ENVELOPE DEBEZIUM` required for anyone that needs to decode Kafka messages produced by Debezium?

**Comment by umanwizard** [comment]
Because for speed, the Debezium envelope processing is heavily entangled with Avro decoding. (Unless someone has gone and made it work with Json since then)

Theoretically it could be extended to support Json on Debezium too but this would require a substantial amount of new code.

**Comment by benesch** [comment]
> @benesch, out of curiosity, why do you think this is the case? Isn't ENVELOPE DEBEZIUM required for anyone that needs to decode Kafka messages produced by Debezium?

Sorry, my question was whether `ENVELOPE DEBEZIUM` was sufficiently composable to be called an an envelope, or whether we ought to be thinking about adding a top-level Debezium source type:

```sql
CREATE SOURCE ... FROM DEBEZIUM 
```

But in further conversations with @petrosagg, I think his position is that it is composable enough to be considered an envelope!

**Comment by petrosagg** [comment]
Debezium envelope is no longer intertwined with our Avro decoding, it operates on a well defined RelationDesc that could have come from decoding JSON or some other source magically outputting that schema. The envelopes only see well typed Row instances in their inputs

**Comment by umanwizard** [comment]
Oh... did you find some way to do this without decoding tons of extra stuff in the `source` metadata stanza? Or did we decide the perf hit is worth it?

**Comment by petrosagg** [comment]
@umanwizard no, I'm just decoding everything. I think the right way to re-introduce this optimization is the same way we do it for the general queries (projection pushdown)

**Comment by aljoscha** [comment]
From an internal thread, on the slightly related `TRANSFORM USING`:

Galaxy-brain idea (that I think I mentioned somewhere before already): Maybe the ingestion pipeline should be fully flexible, and it’s up to users to “build” their ingestion pipeline. It could use something inspired from the Loki/promql syntax (syntax 100% TBD):
```
CREATE SOURCE s
FROM KAFKA CONNECTION k (TOPIC 't')
 | FORMAT JSON
 | ENVELOPE NONE
 | TRANSFORM (SELECT * FROM s)
```

Not sure what other operations there would be, but users could apply them in any order that “typechecks”. Could also use less SQL-y things:

```
CREATE SOURCE s
FROM KAFKA CONNECTION k (TOPIC 't')
 | FORMAT JSON
 | APPLY ENVELOPE NONE
 | FILTER <...>
 | PROJECT <...>
 | MAP <...>
 | FILTER <...> 
```

Also, I’m quite sure we shouldn’t do this! Just throwing it out.
